### PR TITLE
New Raytracing example using OrderedIntersection TreeTraversal

### PR DESCRIPTION
**Pull request by masterleinad**
In #683, we decided to split changing the example from introducing the new traversal.

### CODE REVIEW

**Comment by masterleinad** [comment]
The timings for `./ArborX_RayTracing.exe --Nx=100 --Ny=100 --Nz=100 --rays=10` look pretty much the same as in https://github.com/arborx/ArborX/pull/683#discussion_r901950308
```
BEGIN KOKKOS PROFILING REPORT:
TOTAL TIME: 1.581 seconds
TOP-DOWN TIME TREE:
<average time> <percent of total time> <percent time in Kokkos> <percent MPI imbalance> <remainder> <kernels per second> <number of calls> <name> [type]
=================== 
|-> 1.12e+00 sec 71.1% 58.0% 0.0% 0.5% 1.51e+01 1 intersects_approach [region]
|   |-> 4.90e-01 sec 31.0% 96.4% 0.0% 0.0% 2.65e+01 1 ArborX::query [region]
|   |   |-> 4.90e-01 sec 31.0% 96.4% 0.0% 0.1% 2.65e+01 1 ArborX::CrsGraphWrapper::query::spatial [region]
|   |       |-> 4.82e-01 sec 30.5% 97.8% 0.0% 0.2% 1.87e+01 1 ArborX::CrsGraphWrapper::two_pass [region]
|   |       |   |-> 3.25e-01 sec 20.6% 97.0% 0.0% 3.0% 9.23e+00 1 ArborX::CrsGraphWrapper::two_pass:second_pass [region]
|   |       |   |   |-> 3.15e-01 sec 19.9% 100.0% 0.0% 0.0% 3.17e+00 1 ArborX::BVH::query::spatial [region]
|   |       |   |   |   |-> 3.15e-01 sec 19.9% 100.0% 0.0% ------ 1 ArborX::TreeTraversal::spatial [for]
|   |       |   |-> 1.49e-01 sec 9.4% 100.0% 0.0% 0.0% 6.71e+00 1 ArborX::CrsGraphWrapper::two_pass::first_pass [region]
|   |       |   |   |-> 1.49e-01 sec 9.4% 100.0% 0.0% 0.0% 6.72e+00 1 ArborX::BVH::query::spatial [region]
|   |       |   |       |-> 1.49e-01 sec 9.4% 100.0% 0.0% ------ 1 ArborX::TreeTraversal::spatial [for]
|   |       |   |-> 7.23e-03 sec 0.5% 98.6% 0.0% 1.4% 5.53e+02 1 ArborX::CrsGraphWrapper::first_pass_postprocess [region]
|   |       |   |   |-> 6.86e-03 sec 0.4% 100.0% 0.0% ------ 1 ArborX::Algorithms::exclusive_scan [scan]
|   |       |-> 6.58e-03 sec 0.4% 8.7% 0.0% 91.3% 3.04e+02 1 ArborX::CrsGraphWrapper::query::spatial::compute_permutation [region]
|   |-> 4.55e-01 sec 28.8% 1.4% 0.0% 98.6% 2.20e+00 1 sorting [region]
|   |   |-> 6.18e-03 sec 0.4% 100.0% 0.0% ------ 1 ArborX::Algorithms::iota [for]
|   |-> 1.56e-01 sec 9.9% 100.0% 0.0% ------ 1 deposit_energy [for]
|   |-> 1.73e-02 sec 1.1% 100.0% 0.0% ------ 1 copy sort_array [for]
|-> 4.17e-01 sec 26.4% 98.5% 0.0% 0.1% 1.44e+01 1 ordered_intersects_approach [region]
|   |-> 4.17e-01 sec 26.4% 98.6% 0.0% 0.1% 7.20e+00 1 ArborX::BVH::query::nearest [region]
|   |   |-> 4.10e-01 sec 25.9% 100.0% 0.0% ------ 1 ArborX::Experimental::TreeTraversal::OrderedSpatialPredicate [for]
|   |   |-> 6.18e-03 sec 0.4% 9.4% 0.0% 90.6% 3.24e+02 1 ArborX::BVH::query::nearest::compute_permutation [region]
|-> 9.54e-03 sec 0.6% 53.6% 0.0% 0.2% 6.29e+02 1 problem_setup [region]
|   |-> 8.75e-03 sec 0.6% 56.0% 0.0% 44.0% 5.72e+02 1 make_rays [region]
|   |   |-> 3.47e-03 sec 0.2% 100.0% 0.0% ------ 1 initialize_rays [for]
|-> 4.47e-03 sec 0.3% 36.3% 0.0% 8.2% 1.56e+03 1 ArborX::BVH::BVH [region]
|   |-> 1.81e-03 sec 0.1% 68.7% 0.0% 33.6% 2.21e+03 1 ArborX::BVH::BVH::generate_hierarchy [region]
|   |-> 1.67e-03 sec 0.1% 2.3% 0.0% 97.7% 5.97e+02 1 ArborX::BVH::BVH::sort_linearized_order [region]
```

**Comment by masterleinad** [comment]
Thus, this pull request still includes two versions which I find instructive to keep for comparison.
Also, I decided to move `overlapDistance` back to `src/details/ArborX_Ray.hpp` since it relies on implementation details and seems to fit better there with the already existing version.

File: examples/raytracing/example_raytracing.cpp [hunk 1]
```diff
@@ -119,83 +223,179 @@ int main(int argc, char *argv[])
...
+  // in different order.
+  int n_errors = 0;
+  float rel_tol = 1.e-5;
+  Kokkos::parallel_reduce(
+      "compare", Kokkos::RangePolicy<ExecutionSpace>(exec_space, 0, num_boxes),
+      KOKKOS_LAMBDA(int i, int &error) {
+        using Kokkos::Experimental::fabs;
+        float const rel_error =
+            (energy_intersects(i) == 0.f)
+                ? 0.f
```
Comment (aprokop) [inline examples/raytracing/example_raytracing.cpp:383]: Not sure about this. If `energy_intersects(i)` is 0, and `energy_ordered_intersects(i)` is not, why is the relative error 0?

File: examples/raytracing/example_raytracing.cpp [hunk 2]
```diff
@@ -9,81 +9,179 @@
...
-  KOKKOS_FUNCTION static auto get(Rays<MemorySpace> const &rays, std::size_t i)
-  {
-    return attach(intersects(rays._rays(i)), (int)i);
+    if (l.rid == r.rid)
+      return l.key < r.key;
+    return l.rid < r.rid;
   }
 };
 
+struct IntersectedCell : public IntersectedCellForSorting
```
Comment (aprokop) [inline examples/raytracing/example_raytracing.cpp:128]: If I understand it right, the only reason for inheritance is to save some memory space, so as not to duplicate the partial data. Correct?

**Review by aprokop — COMMENTED** [review]
I mostly reviewed the new code. The question of reconciliation between it and the previous code is still open. @wjge has to weigh on how useful the previous version is.

File: examples/raytracing/example_raytracing.cpp [hunk 3]
```diff
@@ -119,83 +223,179 @@ int main(int argc, char *argv[])
...
+    auto permutation = ArborX::Details::sortObjects(exec_space, sort_array);
+#else
+    Kokkos::View<int *> permutation(
+        Kokkos::view_alloc("permutation", Kokkos::WithoutInitializing),
+        sort_array.size());
+    std::iota(permutation.data(), permutation.data() + sort_array.size(), 0);
+    std::sort(permutation.data(), permutation.data() + sort_array.size(),
+              [&](int const &a, int const &b) {
+                return (sort_array(a) < sort_array(b));
+              });
```
[the remark this answers is not in the document]
Comment (masterleinad) [reply examples/raytracing/example_raytracing.cpp:346]: We can replace this when we have a better idea of the interface to provide for `ArborX::Details::sortObjects`. I don't see that as a reason to hold this pull request.

File: examples/raytracing/example_raytracing.cpp [hunk 4]
```diff
@@ -119,83 +223,179 @@ int main(int argc, char *argv[])
...
-  exec_space.fence();
-  auto time = timer.seconds();
+  // OrderedIntersects-based approach
+  Kokkos::View<float *, MemorySpace> energy_ordered_intersects;
+  {
+    Kokkos::Profiling::pushRegion("ordered_intersects_approach");
+    Kokkos::View<float *, MemorySpace> ray_energy(
+        Kokkos::view_alloc("ray_energy", Kokkos::WithoutInitializing),
+        num_rays * num_boxes);
+    Kokkos::deep_copy(ray_energy, (4000. * dx * dy * dz) / num_rays);
```
[the remark this answers is not in the document]
Comment (masterleinad) [reply examples/raytracing/example_raytracing.cpp:306]: It's a random value. Its value doesn't matter at all.

File: examples/raytracing/example_raytracing.cpp [hunk 5]
```diff
@@ -119,83 +223,179 @@ int main(int argc, char *argv[])
...
+              Kokkos::rand<GeneratorType, float>::draw(random_generator);
+          float xi_4 =
+              Kokkos::rand<GeneratorType, float>::draw(random_generator);
+          float xi_5 =
+              Kokkos::rand<GeneratorType, float>::draw(random_generator);
+
+          float upsilon = 2 * M_PI * xi_1;
+          float theta = acos(1 - 2 * xi_2);
+
+          rays(j + i * num_rays) = {{xi_3 * dx + boxes(i).minCorner()[0],
```
[the remark this answers is not in the document]
Comment (masterleinad) [reply examples/raytracing/example_raytracing.cpp:284]: Sorting predicates still often makes a significant difference in run time. Sometimes it's faster, sometimes it's slower. I don't see playing with this critical for this pull request.

File: examples/raytracing/example_raytracing.cpp [hunk 6]
```diff
@@ -96,19 +194,25 @@ int main(int argc, char *argv[])
...
-    ("L", bpo::value<float>(&L)->default_value(100.0), "size of the domain")
-    ;
-  // clang-format on
+  desc.add_options()("help", "help message")(
+      "rays per box", bpo::value<int>(&num_rays)->default_value(10),
+      "number of rays")("Lx", bpo::value<float>(&Lx)->default_value(1.0),
+                        "Length of X side")(
+      "Ly", bpo::value<float>(&Ly)->default_value(100000.0),
+      "Length of Y side")("Lz", bpo::value<float>(&Lz)->default_value(100000.0),
+                          "Length of Z side")(
```
[the remark this answers is not in the document]
Comment (masterleinad) [reply examples/raytracing/example_raytracing.cpp:212]: Changed it to the ones I used for performance testing now.

File: examples/raytracing/example_raytracing.cpp [hunk 7]
```diff
@@ -119,83 +223,179 @@ int main(int argc, char *argv[])
...
-                        float(std::cos(2 * M_PI * xi_2) * std::sqrt(xi_1)),
-                        float(std::sin(2 * M_PI * xi_2) * std::sqrt(xi_1)),
-                        std::sqrt(1.f - xi_1)}};
+    using RandPoolType = Kokkos::Random_XorShift64_Pool<>;
+    RandPoolType rand_pool(5374857);
+    using GeneratorType = RandPoolType::generator_type;
+
+    Kokkos::parallel_for(
+        "initialize_rays",
+        Kokkos::MDRangePolicy<Kokkos::Rank<2>, ExecutionSpace>(
```
[the remark this answers is not in the document]
Comment (masterleinad) [reply examples/raytracing/example_raytracing.cpp:262]: I haven't noticed that this really shows up when looking at the performance, i.e., this takes 3.47e-03s out of 1.5s run time.

File: examples/raytracing/example_raytracing.cpp [hunk 8]
```diff
@@ -119,83 +223,179 @@ int main(int argc, char *argv[])
...
+                                            num_rays * num_boxes),
+        KOKKOS_LAMBDA(int i) {
+          using Kokkos::Experimental::expm1;
+          float ray_energy = (4000. * dx * dy * dz) / num_rays;
+          for (int j = offsets(i); j < offsets(i + 1); ++j)
+          {
+            float const energy_deposited =
+                -ray_energy *
+                expm1(-values(permutation(j)).optical_path_length);
+            ray_energy += energy_deposited;
```
[the remark this answers is not in the document]
Comment (masterleinad) [reply examples/raytracing/example_raytracing.cpp:364]: I don't agree. This is just two lines and putting the declaration somewhere else is more confusing than helpful. Divergence will be detected by the final test

File: examples/raytracing/example_raytracing.cpp (hunk 1 above)
Comment (masterleinad) [reply examples/raytracing/example_raytracing.cpp:383]: Changed.

File: examples/raytracing/example_raytracing.cpp (hunk 2 above)
Comment (masterleinad) [reply examples/raytracing/example_raytracing.cpp:128]: Yes.

File: examples/raytracing/example_raytracing.cpp (hunk 8 above)
[the remark this answers is not in the document]
Comment (aprokop) [reply examples/raytracing/example_raytracing.cpp:371]: While divergence would indeed be detected, tracking the cause could be hard. It's not a blocker, but I don't consider this resolved.

File: examples/raytracing/example_raytracing.cpp [hunk 9]
```diff
@@ -119,83 +223,179 @@ int main(int argc, char *argv[])
...
+        "initialize_rays",
+        Kokkos::MDRangePolicy<Kokkos::Rank<2>, ExecutionSpace>(
+            exec_space, {0, 0}, {num_boxes, num_rays}),
+        KOKKOS_LAMBDA(const size_t i, const size_t j) {
+          GeneratorType random_generator = rand_pool.get_state();
+          using Kokkos::Experimental::cos;
+          using Kokkos::Experimental::sin;
+          using Kokkos::Experimental::acos;
+
+          float xi_1 =
```
[the remark this answers is not in the document]
Comment (aprokop) [reply examples/raytracing/example_raytracing.cpp:275]: Why was this marked resolved when no change to address this has been made?

File: examples/raytracing/example_raytracing.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (aprokop) [reply examples/raytracing/example_raytracing.cpp]: It is fine if it is not put into the sort utilities file. It would still be helpful if it was made its own function.

File: examples/raytracing/example_raytracing.cpp (hunk 9 above)
[the remark this answers is not in the document]
Comment (masterleinad) [reply examples/raytracing/example_raytracing.cpp:275]: I added a comment above.

File: examples/raytracing/example_raytracing.cpp (hunk 8 above)
[the remark this answers is not in the document]
Comment (masterleinad) [reply examples/raytracing/example_raytracing.cpp]: I introduced a `lost_energy` function now to make progress. Let me know if you think that is better.

**Comment by masterleinad** [comment]
Main questions are:
- Should we have the comparison?
- Should we keep the previous example?

File: examples/raytracing/example_raytracing.cpp [hunk 10]
```diff
@@ -9,81 +9,188 @@
...
+ *     (OrderedIntersectsBased namespace).
+ * 2.) storing all intersections and doing the deposition of energy in a
+ *     postprocessing step (IntersectsBased namespace).
+ */
+
 #include <ArborX.hpp>
 #include <ArborX_Ray.hpp>
-#include <ArborX_Version.hpp>
 
-#include <Kokkos_Core.hpp>
```
Comment (dalg24) [inline examples/raytracing/example_raytracing.cpp:16]: Why did you remove?

File: examples/raytracing/example_raytracing.cpp [hunk 11]
```diff
@@ -9,81 +9,188 @@
...
+#include <Kokkos_Random.hpp>
 
 #include <boost/program_options.hpp>
 
-#include <iostream>
```
Comment (dalg24) [inline examples/raytracing/example_raytracing.cpp:20]: Why did you remove?  You put things into `std::cout`

File: examples/raytracing/example_raytracing.cpp [hunk 12]
```diff
@@ -119,83 +231,182 @@ int main(int argc, char *argv[])
...
+#else
+    Kokkos::View<int *> permutation(
+        Kokkos::view_alloc("Example::permutation", Kokkos::WithoutInitializing),
+        sort_array.size());
+    std::iota(permutation.data(), permutation.data() + sort_array.size(), 0);
+    std::sort(permutation.data(), permutation.data() + sort_array.size(),
+              [&](int const &a, int const &b) {
+                return (sort_array(a) < sort_array(b));
+              });
+#endif
```
Comment (dalg24) [inline examples/raytracing/example_raytracing.cpp:360]: How much does this impact performance?  We are showcasing something we don't want user to do, that is reaching into the `Details::` namespace.
The region label should read "sort[ing] by key".
If the impact on performance is significant I would be ok marking this as a FIXME and using it as a excuse to go ahead and add an abstraction for it in `KokkosExt::` (which we still don't really want users to reach into)

File: examples/raytracing/example_raytracing.cpp [hunk 13]
```diff
@@ -119,83 +231,182 @@ int main(int argc, char *argv[])
...
+                 energy_ordered_intersects(i), energy_intersects(i),
+                 abs_error / fabs(energy_intersects(i)));
+#endif
+          ++error;
+        }
+      },
+      n_errors);
+  std::cout << "errors = " << n_errors << '\n';
+  if (n_errors > 0)
+    Kokkos::abort("The two approaches differ by more than expected!\n");
```
Comment (dalg24) [inline examples/raytracing/example_raytracing.cpp:409]: I would print to `cerr` and return `EXIT_FAILURE`

File: src/geometry/ArborX_Ray.hpp [hunk 14]
```diff
@@ -534,22 +534,36 @@ KOKKOS_INLINE_FUNCTION bool intersection(Ray const &ray, Sphere const &sphere,
...
+}
 
-  // Overlap [tmin, tmax] with [0, +inf)
-  tmin = KokkosExt::max(0.f, tmin);
-
-  // As direction is normalized,
-  //   |(o + tmax*d) - (o + tmin*d)| = tmax - tmin
-  return (tmax - tmin);
+KOKKOS_INLINE_FUNCTION float overlapDistance(Ray const &ray,
+                                             Sphere const &sphere)
```
Comment (dalg24) [inline src/geometry/ArborX_Ray.hpp:561]: Not holding up the PR on this but as I did mention before I am not sure about providing any overload of `overlapDistance` in the library and I would want to reserve the right to remove them later.

File: examples/raytracing/example_raytracing.cpp (hunk 10 above)
Comment (masterleinad) [reply examples/raytracing/example_raytracing.cpp:16]: I just replaced the example. I'm not sure how valuable it is to look at the diff.

File: examples/raytracing/example_raytracing.cpp (hunk 12 above)
Comment (masterleinad) [reply examples/raytracing/example_raytracing.cpp]: Sorting on the host for the benchmark I have been using throughout is around 100x more expensive.

**Review by aprokop — APPROVED** [review]
